### PR TITLE
[llvm] Support IFuncs on Darwin platforms

### PR DESCRIPTION
**Pull request by jroelofs**
... by lowering them as lazy resolve-on-first-use symbol resolvers. Note that
this is subtly different timing than on ELF platforms, where ifunc resolution
happens at load time.

Since ld64 and ld-prime don't support all the cases we need for these, there is
a manual fallback lowering.

### CODE REVIEW

File: llvm/lib/IR/Verifier.cpp [hunk 1]
```diff
@@ -2239,13 +2240,10 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
   }
 
   // Check EVEX512 feature.
-  if (MaxParameterWidth >= 512 && Attrs.hasFnAttr("target-features")) {
-    Triple T(M.getTargetTriple());
-    if (T.isX86()) {
-      StringRef TF = Attrs.getFnAttr("target-features").getValueAsString();
-      Check(!TF.contains("+avx512f") || !TF.contains("-evex512"),
-            "512-bit vector arguments require 'evex512' for AVX512", V);
-    }
```
Comment (jroelofs) [inline llvm/lib/IR/Verifier.cpp:2248]: drive-by fix. We don't need to create a Triple, since we already have one.

File: llvm/test/Verifier/ifunc-macho.ll [hunk 2]
```diff
@@ -0,0 +1,42 @@
+; RUN:  not llvm-as %s -o /dev/null 2>&1 | FileCheck %s
```
Comment (jroelofs) [inline llvm/test/Verifier/ifunc-macho.ll:1]: This test is leftover from a different approach I took earlier in the history of this patch series. I don't think we actually need it.

**Comment by erichkeane** [comment]
So I'm not a good reviewer here, my knowledge of ifuncs comes exclusively from using them to implement multiversioning.

File: llvm/test/CodeGen/AArch64/addrsig-macho.ll [hunk 3]
```diff
@@ -118,8 +118,8 @@ declare void @f3() unnamed_addr
 @a1 = alias i32, i32* @g1
 @a2 = internal local_unnamed_addr alias i32, i32* @g2
 
-@i1 = ifunc void(), void()* ()* @f1
-@i2 = internal local_unnamed_addr ifunc void(), void()* ()* @f2
+@i1 = external ifunc void(), void()* ()* @f1
+@i2 = external local_unnamed_addr ifunc void(), void()* ()* @f2
 
 declare void @llvm.dbg.value(metadata, metadata, metadata)
```
Comment (jroelofs) [inline llvm/test/CodeGen/AArch64/addrsig-macho.ll:125]: I think this change can go away too. It's leftover cruft from a previous version of this patch series.

File: llvm/lib/CodeGen/GlobalISel/CallLowering.cpp [hunk 4]
```diff
@@ -144,7 +144,12 @@ bool CallLowering::lowerCall(MachineIRBuilder &MIRBuilder, const CallBase &CB,
   // Try looking through a bitcast from one function type to another.
   // Commonly happens with calls to objc_msgSend().
   const Value *CalleeV = CB.getCalledOperand()->stripPointerCasts();
-  if (const Function *F = dyn_cast<Function>(CalleeV))
+  if (const GlobalIFunc *IF = dyn_cast<GlobalIFunc>(CalleeV);
+      IF && MF.getTarget().getTargetTriple().isOSBinFormatMachO()) {
+    // ld64 requires that .symbol_resolvers to be called via a stub, so these
+    // must always be a diret call.
```
Comment (aemerson) [inline llvm/lib/CodeGen/GlobalISel/CallLowering.cpp:150]: typo "direct"

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp [hunk 5]
```diff
@@ -1809,6 +1838,292 @@ void AArch64AsmPrinter::emitInstruction(const MachineInstr *MI) {
...
+  else
+    assert(GI.hasLocalLinkage() && "Invalid ifunc linkage");
+
+  OutStreamer->emitCodeAlignment(Align(4), STI);
+  OutStreamer->emitLabel(Name);
+  OutStreamer->emitSymbolAttribute(Name, MCSA_SymbolResolver);
+  emitVisibility(Name, GI.getVisibility());
+
+  // ld-prime does not seem to support aliases of symbol resolvers, so we have
+  // to tail call the resolver manually.
```
Comment (jroelofs) [inline llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1862]: IIUC it is appropriate to mention this here, but not in the actual diff:

rdar://118480251 (Aliased .symbol_resolver's)

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp [hunk 6]
```diff
@@ -1809,6 +1838,292 @@ void AArch64AsmPrinter::emitInstruction(const MachineInstr *MI) {
...
+  // stp	fp, lr, [sp, #-16]!
+  // mov	fp, sp
+  // stp	x1, x0, [sp, #-16]!
+  // stp	x3, x2, [sp, #-16]!
+  // stp	x5, x4, [sp, #-16]!
+  // stp	x7, x6, [sp, #-16]!
+  // stp	d1, d0, [sp, #-16]!
+  // stp	d3, d2, [sp, #-16]!
+  // stp	d5, d4, [sp, #-16]!
+  // stp	d7, d6, [sp, #-16]!
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1977]: Maybe it doesn't matter so much here, but I think it's generally better to adjust SP once and the use immediate offsets vs pre-indexed, since it won't create register dependencies between each store.

File: llvm/test/CodeGen/AArch64/ifunc-asm.ll [hunk 7]
```diff
@@ -0,0 +1,82 @@
+; RUN: llc -mtriple=arm64-unknown-linux-gnu %s -filetype=asm -o - | FileCheck %s --check-prefixes=ELF
+; RUN: llc -mtriple=arm64-apple-darwin %s -filetype=asm -o - -arm64-darwin-ifunc-symbol_resolver=always | FileCheck %s --check-prefixes=MACHO,MACHO-LINKER
+; RUN: llc -mtriple=arm64-apple-darwin %s -filetype=asm -o - -arm64-darwin-ifunc-symbol_resolver=if_supported | FileCheck %s --check-prefixes=MACHO,MACHO-DEFAULT
+; RUN: llc -mtriple=arm64-apple-darwin %s -filetype=asm -o - -arm64-darwin-ifunc-symbol_resolver=never | FileCheck %s --check-prefixes=MACHO,MACHO-MANUAL
+; RUN: llc -mtriple=arm64-apple-darwin %s -filetype=asm -o - | FileCheck %s --check-prefixes=MACHO,MACHO-MANUAL
```
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/ifunc-asm.ll:5]: Needs a RUN line with -global-isel too for end-to-end tests.

**Review by aemerson — COMMENTED** [review]
A few comments so far, I'll probably need to revisit this later.

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp (hunk 6 above)
Comment (jroelofs) [reply llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1977]: Good point. This pretty closely matches what ld64 does, but if we can do better, we should.

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp (hunk 5 above)
Comment (tahonermann) [reply llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1862]: Correct. Per https://llvm.org/docs/DeveloperPolicy.html#commit-messages, it is also ok to include such links in the commit message if you like.

**Comment by jroelofs** [comment]
This is part of a patch series:
    * #73686
    * #73688
    * #73687
    * #73685

**Comment by jroelofs** [comment]
ping

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp [hunk 8]
```diff
@@ -1809,6 +1814,255 @@ void AArch64AsmPrinter::emitInstruction(const MachineInstr *MI) {
   EmitToStreamer(*OutStreamer, TmpInst);
 }
 
+void AArch64AsmPrinter::emitGlobalIFunc(Module &M, const GlobalIFunc &GI) {
+  if (!TM.getTargetTriple().isOSBinFormatMachO())
+    return AsmPrinter::emitGlobalIFunc(M, GI);
```
[the remark this answers is not in the document]
Comment (jroelofs) [reply llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1819]: Sounds good, I like that.

File: llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp [hunk 9]
```diff
@@ -1809,6 +1814,255 @@ void AArch64AsmPrinter::emitInstruction(const MachineInstr *MI) {
...
+  //    *  resolvers cannot have private linkage
+  //    *  resolvers cannot have linkonce linkage
+  //    *  resolvers cannot appear in executables
+  //    *  resolvers cannot appear in bundles
+  //
+  // This works around that by emitting a close approximation of what the linker
+  // would have done.
+
+  auto EmitLinkage = [&](MCSymbol *Sym) {
+    if (GI.hasExternalLinkage() || !MAI->getWeakRefDirective())
```
Comment (MaskRay) [inline llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:1838]: `MAI->getWeakRefDirective()` is always true for Mach-O.

File: llvm/test/CodeGen/AArch64/ifunc-asm.ll [hunk 10]
```diff
@@ -0,0 +1,72 @@
+; RUN: llc -mtriple=arm64-unknown-linux-gnu %s -filetype=asm -o - | FileCheck %s --check-prefixes=ELF
```
Comment (MaskRay) [inline llvm/test/CodeGen/AArch64/ifunc-asm.ll:1]: Omit `-filetype=asm` (convention).

File: llvm/test/CodeGen/X86/ifunc-asm.ll [hunk 11]
```diff
@@ -1,14 +1,49 @@
-; RUN: llvm-as < %s -o - | llc -filetype=asm | FileCheck %s
+; RUN: llc -filetype=asm -mtriple=x86_64-unknown-linux-gnu %s -o - | FileCheck %s --check-prefixes=ELF
```
Comment (MaskRay) [inline llvm/test/CodeGen/X86/ifunc-asm.ll:1]: omit `-filetype=asm`

**Comment by jroelofs** [comment]
ping

File: llvm/include/llvm/CodeGen/AsmPrinter.h [hunk 12]
```diff
@@ -599,6 +599,26 @@ class AsmPrinter : public MachineFunctionPass {
...
+  /// getSubtargetInfo() cannot be used where this is needed because we don't
+  /// have a MachineFunction when we're lowering a GlobalIFunc, and
+  /// getSubtargetInfo requires one. Override the implementation in targets
+  /// that support the Mach-O IFunc lowering.
+  virtual const MCSubtargetInfo *getIFuncMCSubtargetInfo() const {
+    return nullptr;
+  }
+
+  virtual void emitMachOIFuncStubBody(Module &M, const GlobalIFunc &GI,
+                                      MCSymbol *LazyPointer) {
```
Comment (ahmedbougacha) [inline llvm/include/llvm/CodeGen/AsmPrinter.h:611]: Maybe pass the symbol string directly, since the implementations make their own symbol references anyway?  That way you also don't have to do the awkward de-mangle/re-mangle dance with the +1

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 13]
```diff
@@ -2147,24 +2148,80 @@ void AsmPrinter::emitGlobalIFunc(Module &M, const GlobalIFunc &GI) {
...
+    OutStreamer->emitSymbolAttribute(Name, MCSA_ELF_TypeIndFunction);
+    emitVisibility(Name, GI.getVisibility());
+
+    // Emit the directives as assignments aka .set:
+    const MCExpr *Expr = lowerConstant(GI.getResolver());
+    OutStreamer->emitAssignment(Name, Expr);
+    MCSymbol *LocalAlias = getSymbolPreferLocal(GI);
+    if (LocalAlias != Name)
+      OutStreamer->emitAssignment(LocalAlias, Expr);
+  } else if (TM.getTargetTriple().isOSBinFormatMachO() &&
```
Comment (ahmedbougacha) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:2172]: early return with if->fatal_error, for indentation?

**Review by ahmedbougacha — COMMENTED** [review]
Only a couple minor comments, LG otherwise, thanks!  The Subtarget weirdness is unfortunate but I can't say I have a better alternative

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 14]
```diff
@@ -2169,8 +2169,11 @@ void AsmPrinter::emitGlobalIFunc(Module &M, const GlobalIFunc &GI) {
     MCSymbol *LocalAlias = getSymbolPreferLocal(GI);
     if (LocalAlias != Name)
       OutStreamer->emitAssignment(LocalAlias, Expr);
-  } else if (TM.getTargetTriple().isOSBinFormatMachO() &&
-             getIFuncMCSubtargetInfo()) {
+
+    return;
+  }
+
+  if (TM.getTargetTriple().isOSBinFormatMachO() && getIFuncMCSubtargetInfo()) {
```
Comment (ahmedbougacha) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:2176]: oh I meant literally:
```
  if (!TM.getTargetTriple().isOSBinFormatMachO() || !getIFuncMCSubtargetInfo())
      llvm::report_fatal_error("IFuncs are not supported on this platform");
```

Comment (jroelofs) [reply llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:2176]: oh, I see. sure.

**Review by ahmedbougacha — APPROVED** [review]
Neat, ty!
